networkit/cpp/embedding/AliasSampler.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * AliasSampler.cpp
+ *
+ *  Created on: 03.07.2020
+ *      Author: Klaus Ahrens  <ahrens@informatik.hu-berlin.de>
+ *
+ *  adapted and reimplemented from node2vec
+ *  part of snap [https://github.com/snap-stanford/snap]
+ *  Copyright (c) 2007-2019, Jure Leskovec (under BSD license)
+ *
+ *  see [https://arxiv.org/pdf/1607.00653v1.pdf]
+ *
+ */
+
+// networkit-format
+
+#include <random>
+#include <vector>
+
+#include <networkit/Globals.hpp>
+#include <networkit/auxiliary/Random.hpp>
+
+#include "AliasSampler.hpp"
+
+namespace NetworKit {
+namespace Embedding {
+
+double uniform_real() {
+    static std::uniform_real_distribution<> dist;
+
+    return dist(Aux::Random::getURNG());
+}
```

Comment:
What is the advantage of this function over `Aux::Random::real()`?

Reply:
As reported in the docs [1], if that code is be used in "hot" sections, it might be better to keep a local `uniform_real_distribution` and avoid to use `Aux::Random::real()`.
[1] https://networkit.github.io/dev-docs/cpp_api/function_namespaceAux_1_1Random_1ab0532015f7ed7d0eddffeb10405720f9.html#exhale-function-namespaceaux-1-1random-1ab0532015f7ed7d0eddffeb10405720f9

Reply:
Even though it is not completely clear, I assume that the slowdown mentioned in the documentation is compared to a) holding a local (!) reference to the URNG and b) a local distribution object. I assume that already calling `getURNG()` introduces most of the overhead. If this overhead shall be avoided, I would suggest adding a reference to the URNG and the distribution object in the `AliasSampler` class. Note that neither the URNG nor the `uniform_real_distribution` are thread-safe, but again this implementation is already not thread-safe, so I assume that this is only called from a single thread anyways?

Reply:
The implementation should be made thread-safe by removing global state.

Reply:
Taking a reference to getURNG() makes sense to avoid the overhead of the thread local access (but it should be stored in a member variable). The distribution object should also be re-used and not be recreated.